kdbg/debugger.cpp

```diff
@@ -2151,17 +2153,9 @@ void KDebugger::slotDisassemble(const QString& fileName, int lineNo)
     }
 }
 
-void KDebugger::setDisassemblyFlavor(QString flavor)
+void KDebugger::submitDisassemblyFlavor()
 {
-    // DefaultGenericFlavor -> "default"
-    if (flavor == DefaultGenericFlavor) {
-        flavor = m_globalFlavor;
-    }
-
-    if(m_tmpFlavor != flavor){
-	m_tmpFlavor = flavor;
-	m_d->executeCmd(DCsetdisassflavor, flavor);
-    }
+    m_d->executeCmd(DCsetdisassflavor, m_flavor);
```

Comment:
That is actually not how I thought it to be. You should keep all the logic here that decides which flavor to choose depending on which option says "default, please", but you should not change the state (`m_flavor`) here. In contrast, everywhere else you only change the state, but do not have any logic whether there is a default option that must be overridden.

- In `setDefaultFlavor()` you only set `m_globalFlavor` without applying any logic.
- In `programSettings()` and `restoreProgramSettings()` you set `m_flavor`, also without any logic.
- In `saveProgramSettings()` you store `m_flavor` without applying any logic.

That is, you can have all 9 combinations of { default, att, intel } x { default, att, intel } in `m_flavor` and `m_globalFlavor` and you decide here, which one to submit.

Reply:
I am afraid I am confused. `setDefaultFlavor` can set `m_globalFlavor` unconditionally. But how `programSettings()` and `restoreProgramSettings()` are going to set m_flavor unconditionally and to what value?

Also, `m_flavor` isn't saved or stored anywhere, it is created and destroyed every time a program is loaded and closed. Unless I misunderstood something.

Reply:
> But how `programSettings()` and `restoreProgramSettings()` are going to set m_flavor unconditionally and to what value?

To the value that the user chooses in the *This Program* settings. This is a completely independent setting from the global option. And then `submitDisassemblyFlavor()` is the only place that weaves the two settings into the setting that is submitted to Gdb.

> Also, `m_flavor` isn't saved or stored anywhere, it is created and destroyed every time a program is loaded and closed.

That is an omission that must be fixed. (I noticed this only today.) The places to save and restore this setting are `KDebugger::saveProgramSettings()` and `KDebugger::restoreProgramSettings()`. I suggest to put it into the `GeneralGroup`.

Reply:
> To the value that the user chooses in the This Program settings. This is a completely independent setting from the global option. And then submitDisassemblyFlavor() is the only place that weaves the two settings into the setting that is submitted to Gdb. 

But `m_flavor` will already be reseted from the _This Program_ dialog once it's open. `programSettings()` and `restoreProgramSettings()` will have nothing to set `m_flavor` to. Unless, we allow the logic (if-else etc) to read `m_globalFlavor` in one of these functions and set `m_flavor` accordingly.

> That is an omission that must be fixed. (I noticed this only today.) The places to save and restore this setting are KDebugger::saveProgramSettings() and KDebugger::restoreProgramSettings(). I suggest to put it into the GeneralGroup.

I am not so sure about this and it's not from a technical point of view. Each time a `KDebugger` instance is constructed, by definition `m_flavor` initially (at least) obeys `m_globalFlavor` unless the user explicitly chooses to override it. I fail to see why a "temporary exception" from a general rule (global flavor) should be preserved.

Reply:
> I am not so sure about this and it's not from a technical point of view. Each time a `KDebugger` instance is constructed, by definition `m_flavor` initially (at least) obeys `m_globalFlavor` unless the user explicitly chooses to override it. I fail to see why a "temporary exception" from a general rule (global flavor) should be preserved.

But the options that can be set in *This Program* are not temporary; they must be persistent. If you think that this does not make sense, then it is better not to allow to set the disassembly flavor in *This Program* and only use the global option.

Think of it like this: If it is difficult to explain to a user how an option or a combination of options works, then there is room for improvement in the design. Perhaps it's best to start with only one of the two option settings, and then later when we know what we need and how we want it to work, we add the other?

Reply:
But why isn't temporary? Next time a program will be loaded, global flavor is going to be used again and m_flavor in Kdebugger will be scratched from it's previous value.

Reply:
Are we on the same page what the *This Program* settings actually do? Those settings are stored in a configuration file that applies to "this particular program". For example, the breakpoints are stored there, because they are not "temporary". It does not make sense to store breakpoints globally, but does make sense that the breakpoints are restored when you load the program again. The same should happen when you set the flavor option in *This Program*. There, the user can choose "att", "intel", "use global option", and can express two kinds of wishes: (1) "Always use this flavor for this program (att or intel)" and (2) "I don't particularly care which flavor to use for this program, just use whatever I have set in the global options". And that's already the explanation that the user needs.

> Do you want me to keep only one global flavor and apply it immediately in kdbg?

That would be a good start, because it covers the majority of use cases: I would guess that most people usually don't switch architectures a lot, but they do care about the assembly flavor.

The *This Program* setting entered the game only because we thought about catering to more complicated use cases. We can postpone that.

I'm not 100% sure what you mean by "apply it immediately". You should not necessarily change the times when the option is submitted to Gdb, because you should keep an eye on how things will have to work when the *This Program* option comes back. (We should not be forced to rewrite new code when we already know how the end state is going to look like.)

Reply:
Ok, now I understand about the first part. First of all, now that I understand I agree with your concept. But how should I store m_flavor for each program the user will load in years perhaps? They can be hundreds.

Second, what I mean immediately is that the flavor set in global settings, will update the view as soon as the OK button is pressed, without needing to restart. 

This will indeed interfere with the current code especially if this program will ever support flavour settings. So my ideas are that either I make this program dialog do what you want (with plenty of your help if you can) or we leave the flavor to be controlled globally which is simpler but perhaps frustrating users.

While the simpler, global option is easier for me (and maybe easier to maintain), I vote for feature completeness. But again, I am quite lost here with all the logic in this function but not the other etc. It's up to you 👍

Reply:
> But how should I store m_flavor for each program the user will load in years perhaps? They can be hundreds.

There's already code to do that. Have a look at `KDebugger::saveProgramSettings()` and `KDebugger::restoreProgramSettings()`. You can just add another setting in the `GeneralGroup`.

> Second, what I mean immediately is that the flavor set in global settings, will update the view as soon as the OK button is pressed, without needing to restart.

I agree that this should happen. Just make it so that `submitDisassFlavor()` is invoked. It will trigger the update of the display from its `handleSetDisassFlavor()`.

I'm OK that we start with just the global option, if you are more comfortable and leave out the *This Program* setting for now.

Reply:
`submitDisassFlavor()` actually will take a qstring argument if dbgmainwindow calls it.

I will try to finish this feature without leaving any loose ends for later.

Reply:
I have to ask. Why is it necessary to overwrite `m_flavor` in `submitDisassemblyFlavor()`? In other words, why does this function not look like this?

```
    QString flavor = m_flavor;

    // fall back to global setting if nothing specific requested
    if (flavor.isEmpty() || flavor == WhatEverTheDefaultIsIfItIsNotJustTheEmptyString)
        flavor = m_globalFlavor;

    // do nothing if still no specific setting requested
    if (!flavor.isEmpty() && flavor != WhatEverTheDefaultIsIfItIsNotJustTheEmptyString)
        m_d->executeCmd(DCsetdisassflavor, flavor);
```

A secondary question is what `WhatEverTheDefaultIsIfItIsNotJustTheEmptyString` should be. You use `default` for the global setting and `use global setting` for the *This Program* setting. I strongly suggest that we use `default` for both. There is no reason to have different strings that both mean "default". (I would even just use the empty string, but you don't seem to be comfortable with that, but never mind.)

The way you have written the option entry in the *This Program* dialog so far, the option text would then also read "default" without letting the user know that its meaning is "use global setting". But that is just a matter of presentation and can be fixed later.